TrimZero.lua

```diff
@@ -0,0 +1,139 @@
+------------------------------------------------------------------------
+--[[ TrimZero ]]--
+-- Author: Jin-Hwa Kim
+-- License: LICENSE.2nd.txt
+
+-- Decorator that zeroes the output rows of the encapsulated module
+-- for commensurate input rows which are tensors of zeros
+
+-- The only difference from `MaskZero` is that it reduces computational costs 
+-- by varying a batch size, if any, for the case that varying lengths 
+-- are provided in the input. Notice that when the lengths are consistent, 
+-- `MaskZero` will be faster, because `TrimZero` has an operational cost. 
+
+-- In short, the result is the same with `MaskZero`'s, however, `TrimZero` is
+-- faster than `MaskZero` only when sentence lengths is costly vary.
+-- In practice, e.g. language model, `TrimZero` is expected to be faster than
+--  `MaskZero` 30%. (You can test with it using `test/test_trimzero.lua`.)
+
+-- Zero vectors (i.e. padding) must be at the beginning of the sequence
+-- because this decorator will otherwise reset the recurrentModule
+-- in the middle or after the sequence
+-- TODO add assertion in case padding in uncountered after non padding ?
+------------------------------------------------------------------------
+local TrimZero, parent = torch.class("nn.TrimZero", "nn.MaskZero")
+
+require 'torchx'
+
+function TrimZero:__init(module, nInputDim, silent)
+   parent.__init(self, module, nInputDim, silent)
+end
+
+function TrimZero:recursiveMask(output, input, mask)
+   if torch.type(input) == 'table' then
+      output = torch.type(output) == 'table' and output or {}
+      for k,v in ipairs(input) do
+         output[k] = self:recursiveMask(output[k], v, mask)
+      end
+   else
+      assert(torch.isTensor(input))
+      output = torch.isTensor(output) and output or input.new()
+      
+      -- make sure mask has the same dimenion as the input tensor
+      local inputSize = input:size():fill(1)
+      if input:dim() - 1 == self.nInputDim then
+         inputSize[1] = input:size(1)
+      end
+      mask:resize(inputSize)
+      -- build mask
+      if input:dim() - 1 == self.nInputDim then
+         output:resizeAs(input):copy(input)
```

Comment:
Could you use a lazily allocated buffer here to reduce memory allocations in the following `output = output:index...`. So instead we would have `output:index(buffer, ....`